hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/DataNodeMetricsService.java

```diff
@@ -0,0 +1,342 @@
+          LOG.warn("Collection polling interrupted");
+          Thread.currentThread().interrupt();
+          break;
+        }
+      }
+    }
+  }
+
+  private boolean checkAndHandleTimeout(
+      DatanodePendingDeletionMetrics key, Future<DatanodePendingDeletionMetrics> future,
+      CollectionContext context, long currentTime) {
+    long elapsedTime = currentTime - context.submissionTimes.get(key);
+    if (elapsedTime > PER_NODE_TIMEOUT_MS && !future.isDone()) {
+      LOG.warn("Task for datanode {} [{}] timed out after {}ms",
+          key.getHostName(), key.getDatanodeUuid(), elapsedTime);
+      future.cancel(true); // Interrupt the task
+      context.failed++;
+      context.results.add(key); // Add with -1 (failed)
+      return true;
+    }
+    return false;
+  }
+
+  private void handleCompletedFuture(
+      DatanodePendingDeletionMetrics key, Future<DatanodePendingDeletionMetrics> future,
+      CollectionContext context) {
+    try {
+      DatanodePendingDeletionMetrics result = future.get();
+      if (result.getPendingBlockSize() < 0) {
+        context.failed++;
+      } else {
+        context.totalPending += result.getPendingBlockSize();
+      }
+      context.results.add(result);
+      LOG.debug("Processed result from {}", key.getHostName());
+    } catch (ExecutionException | InterruptedException e) {
+      String errorType = e instanceof InterruptedException ? "interrupted" : "execution failed";
+      LOG.error("Task {} for datanode {} [{}]: {}",
+          errorType, key.getHostName(), key.getDatanodeUuid(), e.getMessage());
+      context.failed++;
+      context.results.add(key);
+      if (e instanceof InterruptedException) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Atomically updates the class's shared state with the results from the collection context.
+   */
+  private void updateFinalState(CollectionContext context) {
+    // Update shared state atomically
+    synchronized (this) {
+      pendingDeletionList = context.results;
+      totalPendingDeletion = context.totalPending;
+      totalNodesQueried = context.totalQueried;
+      totalNodesFailed = context.failed;
+      currentStatus = MetricCollectionStatus.SUCCEEDED;
+      lastCollectionEndTime = System.currentTimeMillis();
+    }
+
+    LOG.info("Metrics collection completed. Queried: {}, Failed: {}",
+        context.totalQueried, context.failed);
+  }
+
+  /**
+   * Resets the collection state.
+   */
+  private void resetState() {
+    pendingDeletionList = new ArrayList<>();
+    totalPendingDeletion = 0L;
+    totalNodesQueried = 0;
+    totalNodesFailed = 0;
+  }
+
+  public synchronized DataNodeMetricsServiceResponse getCollectedMetrics() {
+    if (currentStatus != MetricCollectionStatus.SUCCEEDED) {
+      return DataNodeMetricsServiceResponse.newBuilder()
+          .setStatus(currentStatus)
+          .build();
+    }
+    return DataNodeMetricsServiceResponse.newBuilder()
```

Comment:
Please explicitly check currentStatus agaist expected MetricCollectionStatus, and return the data, instead of a !SUCCEEDED check, so if there is race condition issue due to concurrent calls, it will be caught. 

Same for handleDataNodeMetrics, please check status is inprogress.